examples/custom-node/src/evm/assembler.rs

```diff
@@ -1,35 +1,118 @@
-use crate::chainspec::CustomChainSpec;
-use alloy_consensus::Block;
-use alloy_evm::block::{BlockExecutionError, BlockExecutorFactory};
+use crate::{
+    chainspec::CustomChainSpec,
+    primitives::{Block, BlockBody, CustomHeader, CustomTransaction},
+};
+use alloy_consensus::{
+    constants::EMPTY_WITHDRAWALS, proofs, Header, TxReceipt, EMPTY_OMMER_ROOT_HASH,
+};
+use alloy_eips::{eip7685::EMPTY_REQUESTS_HASH, merge::BEACON_NONCE};
+use alloy_evm::block::{BlockExecutionError, BlockExecutionResult, BlockExecutorFactory};
 use alloy_op_evm::OpBlockExecutionCtx;
+use alloy_primitives::logs_bloom;
 use reth_ethereum::{
     evm::primitives::execute::{BlockAssembler, BlockAssemblerInput},
     primitives::Receipt,
 };
-use reth_op::{node::OpBlockAssembler, DepositReceipt, OpTransactionSigned};
+use reth_op::DepositReceipt;
+use reth_optimism_consensus::{calculate_receipt_root_no_memo_optimism, isthmus};
+use reth_optimism_forks::OpHardforks;
 
 #[derive(Clone, Debug)]
 pub struct CustomBlockAssembler {
-    inner: OpBlockAssembler<CustomChainSpec>,
+    chain_spec: CustomChainSpec,
 }
 
 impl<F> BlockAssembler<F> for CustomBlockAssembler
 where
     F: for<'a> BlockExecutorFactory<
         ExecutionCtx<'a> = OpBlockExecutionCtx,
-        Transaction = OpTransactionSigned,
+        Transaction = CustomTransaction,
         Receipt: Receipt + DepositReceipt,
     >,
 {
-    // TODO: use custom block here
-    type Block = Block<OpTransactionSigned>;
+    type Block = Block;
 
     fn assemble_block(
         &self,
-        input: BlockAssemblerInput<'_, '_, F>,
+        input: BlockAssemblerInput<'_, '_, F, CustomHeader>,
     ) -> Result<Self::Block, BlockExecutionError> {
-        let block = self.inner.assemble_block(input)?;
+        let BlockAssemblerInput {
+            evm_env,
+            execution_ctx: ctx,
+            transactions,
+            output: BlockExecutionResult { receipts, gas_used, .. },
+            bundle_state,
+            state_root,
+            state_provider,
+            ..
+        } = input;
+
+        let timestamp = evm_env.block_env.timestamp;
+
+        let transactions_root = proofs::calculate_transaction_root(&transactions);
+        let receipts_root =
+            calculate_receipt_root_no_memo_optimism(receipts, &self.chain_spec, timestamp);
+        let logs_bloom = logs_bloom(receipts.iter().flat_map(|r| r.logs()));
+
```

Comment:
I feel like for all of this we could use helpers,

maybe even on BlockAssemblerInput itself, for

tx_root, logs_bloom, receiptsroot